src/java/org/apache/cassandra/db/compaction/ShardManager.java

```diff
@@ -286,43 +284,50 @@ private static <T, R extends CompactionSSTable> List<T> applyMaxParallelism(int
             return tasks;
         }
 
-        // Otherwise we have to group shards together. Define a target token span per task and greedily group
-        // to be as close to it as possible.
-        double spanPerTask = shards.stream().map(Pair::right).mapToDouble(t -> t.left.size(t.right)).sum() / maxParallelism;
-        double currentSpan = 0;
+        double totalSpan = shards.stream().map(Pair::right).mapToDouble(r -> r.left.size(r.right)).sum();
+        double spanPerTask = totalSpan / maxParallelism;
+
+        List<T> tasks = new ArrayList<>();
         Set<R> currentSSTables = new HashSet<>();
         Token rangeStart = null;
-        Token prevEnd = null;
-        List<T> tasks = new ArrayList<>(maxParallelism);
-        for (var pair : shards)
+        double currentSpan = 0;
+        int shardsRemaining = shards.size();
+        int tasksRemaining = maxParallelism;
+
+        for (Pair<Set<R>, Range<Token>> pair : shards)
         {
-            final Token currentEnd = pair.right.right;
-            final Token currentStart = pair.right.left;
+            Token currentStart = pair.right.left;
+            Token currentEnd = pair.right.right;
             double span = currentStart.size(currentEnd);
+
             if (rangeStart == null)
                 rangeStart = currentStart;
-            if (currentSpan + span >= spanPerTask - 0.001) // rounding error safety
+
+            currentSSTables.addAll(pair.left);
+            currentSpan += span;
+            shardsRemaining--;
+
+            boolean isLastTask = tasksRemaining == 1;
+            boolean shouldEmit = !isLastTask &&
```

Comment:
I find it a little difficult to understand what this does and why. I took a stab at rearranging the loop [here](https://github.com/datastax/cassandra/commit/808a37ab5f79c278e712dcb6c908fc46b846a988); let me know which version you prefer to go with.